File: test/VbNet.LanguageServer.Tests/Services/FoldingRangeServiceTests.cs
using Microsoft.Extensions.Logging.Abstractions;
using VbNet.LanguageServer.Protocol;
using VbNet.LanguageServer.Services;
using VbNet.LanguageServer.Workspace;
using Xunit;

namespace VbNet.LanguageServer.Tests.Services;

/// <summary>
/// Unit tests for FoldingRangeService.
/// </summary>
public class FoldingRangeServiceTests
{
    private readonly WorkspaceManager _workspaceManager;
    private readonly DocumentManager _documentManager;
    private readonly FoldingRangeService _foldingRangeService;

    public FoldingRangeServiceTests()
    {
        _workspaceManager = new WorkspaceManager(NullLogger<WorkspaceManager>.Instance);
        _documentManager = new DocumentManager(_workspaceManager, NullLogger<DocumentManager>.Instance);
        _foldingRangeService = new FoldingRangeService(
            _workspaceManager,
            _documentManager,
            NullLogger<FoldingRangeService>.Instance);

        _workspaceManager.Initialize();
    }

    [Fact]
    public async Task GetFoldingRangesAsync_NoDocument_ReturnsEmpty()
    {
        var @params = new FoldingRangeParams
        {
            TextDocument = new TextDocumentIdentifier { Uri = "file:///nonexistent.vb" }
        };

        var result = await _foldingRangeService.GetFoldingRangesAsync(@params, CancellationToken.None);

        Assert.NotNull(result);
        Assert.Empty(result);
    }

    [Fact]
    public async Task GetFoldingRangesAsync_StandaloneDocument_ReturnsRegion()
    {
        var uri = "file:///c:/test/module.vb";
        var text = """
                   #Region "Test"
                   Module Module1
                       Sub Main()
                           Dim x = 1
                       End Sub
                   End Module
                   #End Region
                   """;

        _documentManager.HandleDidOpen(new DidOpenTextDocumentParams
        {
            TextDocument = new TextDocumentItem
            {
                Uri = uri,
                LanguageId = "vb",
                Version = 1,
                Text = text
            }
        });

        var @params = new FoldingRangeParams
        {
            TextDocument = new TextDocumentIdentifier { Uri = uri }
        };

        var result = await _foldingRangeService.GetFoldingRangesAsync(@params, CancellationToken.None);

        Assert.NotNull(result);
        Assert.Contains(result, range => range.Kind == FoldingRangeKind.Region);
    }
}

File: test/VbNet.LanguageServer.Tests.Vb/Services/FoldingRangeServiceTests.vb
Imports System.Threading
Imports System.Threading.Tasks
Imports Microsoft.Extensions.Logging.Abstractions
Imports VbNet.LanguageServer.Protocol
Imports VbNet.LanguageServer.Services
Imports VbNet.LanguageServer.Workspace
Imports Xunit

Namespace VbNet.LanguageServer.Tests.Services

    ''' <summary>
    ''' Unit tests for FoldingRangeService.
    ''' </summary>
    Public Class FoldingRangeServiceTests
        Private ReadOnly _workspaceManager As WorkspaceManager
        Private ReadOnly _documentManager As DocumentManager
        Private ReadOnly _foldingRangeService As FoldingRangeService

        Public Sub New()
            _workspaceManager = New WorkspaceManager(NullLogger(Of WorkspaceManager).Instance)
            _documentManager = New DocumentManager(_workspaceManager, NullLogger(Of DocumentManager).Instance)
            _foldingRangeService = New FoldingRangeService(
                _workspaceManager,
                _documentManager,
                NullLogger(Of FoldingRangeService).Instance)

            _workspaceManager.Initialize()
        End Sub

        <Fact>
        Public Async Function GetFoldingRangesAsync_NoDocument_ReturnsEmpty() As Task
            Dim request = New FoldingRangeParams With {
                .TextDocument = New TextDocumentIdentifier With {.Uri = "file:///nonexistent.vb"}
            }

            Dim result = Await _foldingRangeService.GetFoldingRangesAsync(request, CancellationToken.None)

            Assert.NotNull(result)
            Assert.Empty(result)
        End Function

        <Fact>
        Public Async Function GetFoldingRangesAsync_StandaloneDocument_ReturnsRegion() As Task
            Dim uri = "file:///c:/test/module.vb"
            Dim text = String.Join(vbLf,
                "#Region ""Test""",
                "Module Module1",
                "    Sub Main()",
                "        Dim x = 1",
                "    End Sub",
                "End Module",
                "#End Region")

            _documentManager.HandleDidOpen(New DidOpenTextDocumentParams With {
                .TextDocument = New TextDocumentItem With {
                    .Uri = uri,
                    .LanguageId = "vb",
                    .Version = 1,
                    .Text = text
                }
            })

            Dim request = New FoldingRangeParams With {
                .TextDocument = New TextDocumentIdentifier With {.Uri = uri}
            }

            Dim result = Await _foldingRangeService.GetFoldingRangesAsync(request, CancellationToken.None)

            Assert.NotNull(result)
            Assert.Contains(result, Function(range) range.Kind = FoldingRangeKind.Region)
        End Function
    End Class

End Namespace
